feat(api): add Thermocycler Gen2 driver

fsinapi opened the pull request:
# Overview

Makes changes to the Thermocycler hardware controller & driver in order to allow control of Thermocycler Gen2 modules. When loading a Thermocycler object, a driver-level factory function detects the communication protocol in use and distinguishes between the two generations of Thermocycler module.

Tested on a robot through the Jupyter Notebook with a Thermocycler Gen2 connected. Was able to open/close the lid, set temperataure profiles, and generally control it using the normal Thermocycler API. I modified the VID and PID on the firmware to match the old thermocycler to get it to enumerate correctly, the new USB symlink association for `udev` still needs to be added.

# Changelog

- Added a `ThermocyclerDriverV2` class which implements any communication protocol changes for the Gen2 thermocycler.
- Added a `ThermocyclerDriverFactory` static class which sends the Get Device Info command to a thermocycler and parses the response to determine whether it is a Gen1 or Gen2 module, and initializes the correct instance of the driver.
- Made the `bootloader` function of the hardware controller modules a member method rather than a class method, because the bootloader type of a Thermocycler depends on which generation the instance is.
- Added dummy json definitions for the Thermocycler Gen2. These are just copied directly from the TC Gen1.
- Added functions to change the ACK and ERROR keywords on a serial connection after initialization. Used to adapt the serial connection for a Gen1 or Gen2 TC.
- Modified Thermocycler Simulator driver to be able to store its model info
- Updated ThermocyclerMovementFlagger to look for Gen1 _and_ Gen2 thermocyclers

# Review requests

- Does it make sense to have the dummy .json definitions for the TC Gen2? Should those just be removed until we have a real hardware definition?
- I've only tested by running simple scripts through Jupyter, are there other locations in the software stack that might be affected that this testing wouldn't cover?
  - **EDIT** tested on a local emulator by pushing a Python protocol and executing it with a run, was able to run fine.

# Risk assessment

Medium? With the exception of adding the Model enumeration, this should only affect hardware control-level code, but I am unfamiliar enough with the rest of the codebase that there may be side effects I'm overlooking

fsinapi committed:
Added utility functions to serial_connection.py to modify the ack and error keywords to support different TC drivers

fsinapi committed:
- bootloader() is no longer a class function, because on the Thermocycler hardware controller it depends on what thermocycler is actually connected
- Added a ThermocyclerDriverV2 which implements any communication protocol differences from the Gen1 thermocycler driver. A factory driver is used to generate the correct version of the driver, based off of the return format of the connected thermocycler's M115 response
- Added enumerated type for ThermocyclerModuleV2

fsinapi committed:
Added a dummy ThermocyclerModule definition

fsinapi committed:
Thermocycler simulator now utilizes the sim_model

fsinapi committed:
Added Thermocycler V2 model enumerations where V1 was present

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

# Conflicts:
#	api/tests/opentrons/protocol_engine/conftest.py

fsinapi committed:
Added a dummy definitions/3/thermocyclerModuleV2 copied from the V1

fsinapi committed:
Formatting

fsinapi committed:
lint

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

sfoster1 commented on api/src/opentrons/drivers/thermocycler/driver.py:
```diff
@@ -69,12 +73,43 @@ async def create(
+        else:
+            connection.set_ack(TC_ACK)
+            # Must cycle the connection because the old setting of the ACK did
+            # not capture the entire response.
```
can we [`reset_input_buffer`](https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.reset_input_buffer) instead?

sfoster1 commented on api/src/opentrons/drivers/thermocycler/driver.py:
```diff
@@ -233,3 +268,43 @@ async def enter_programming_mode(self) -> None:
+    async def set_ramp_rate(self, ramp_rate: float) -> None:
+        """Send a set ramp rate command"""
+        # This command is fully unsupported on TC Gen2
+        return None
```
probably want to `raise NotImplementedError()` instead

fsinapi replied:
Yeah this one is really weird. My thinking on having it silently return is because that's basically what happens with the Gen1 - the firmware will accept a Set Ramp Rate GCode, but nothing actually happens. Users can specify a ramp rate and it won't throw any exceptions with a Gen1, so that's my only apprehension.

sfoster1 commented on shared-data/module/definitions/2/thermocyclerModuleV2.json:
```diff
@@ -0,0 +1,23 @@
+    "x": 14.4,
+    "y": 64.93
+  },
+  "displayName": "Thermocycler Module",
```
```suggestion
  "displayName": "Thermocycler Module GEN2",
```

sfoster1 reviewed (approved):
Driver side LGTM, couple little things I think should change but overall solid. Re: the dummy module def - We definitely should have it just to establish the space it will fill.

@mcous @b-cooper is there anywhere we have to add this to a denylist to avoid showing up in the app or PD?

fsinapi committed:
linter linter linter

fsinapi committed:
Update shared-data/module/definitions/2/thermocyclerModuleV2.json

Co-authored-by: Seth Foster <seth@opentrons.com>

fsinapi committed:
Added a `rest_input_buffer` function to the AsyncSerial and SerialConnection classes, and use that instead of cycling connection.

amitlissack commented on api/src/opentrons/drivers/thermocycler/driver.py:
```diff
@@ -50,17 +49,22 @@ class GCODE(str, Enum):
 
         Args:
-            port: port or url of temp deck
+            port: port or url of thermocycler
```
I wonder if a different approach could be "neater". 

`SerialConnection` has a constructor that takes an `AsyncSerial` argument. You don't have to use `create`.

Perhaps you can create an `AsyncSerial` object and hold on to it. You create a `SerialConnection` (using its constructor) to check for Thermocycler Version. If you learn that the `SerialConnection` does not match the thermocycler version, you create a different one with the different ack value. 

Doing that free you from adding the `set_ack` and `set_error` methods to `SerialConnection`. Even `reset_input_buffer` won't be necessary on `SerialConnection`. 

In pseudo code:

```
serial_port = await AsyncSerial.create(
            port=port,
            baud_rate=baud_rate,
            timeout=timeout,
            loop=loop,
            reset_buffer_before_write=reset_buffer_before_write,
        )

connection = SerialConnetion(serial=serial_port, ack=gen2_ack, ....)

if not self.is_gen2(connection): 
     serial_port.reset_input_buffer()
    return SerialConnetion(serial=serial_port, ack=gen1_ack, ....)
else:
    return connection
```

What do you think?

fsinapi replied:
Sounds like a much cleaner approach!

amitlissack commented on api/src/opentrons/drivers/thermocycler/driver.py:
```diff
@@ -69,12 +73,42 @@ async def create(
+            return ThermocyclerDriver(connection)
+
+    @staticmethod
+    async def check_for_gen2_protocol(connection: SerialConnection) -> bool:
```
Could we rename this to `is_gen2_protocol`?

amitlissack reviewed (commented):
Just a thought about avoiding some changes to `SerialConnection`.

fsinapi committed:
Added Thermocycler Gen2 model enumeration in a couple places it was left out

fsinapi committed:
Some cleanup per Amit's review

amitlissack reviewed (approved):
Thanks for that 👍  . Happy it worked out!

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

fsinapi committed:
Fix linter error

mcous commented on api/src/opentrons/protocol_engine/types.py:
```diff
@@ -145,6 +145,7 @@ class ModuleModel(str, Enum):
     MAGNETIC_MODULE_V1 = "magneticModuleV1"
     MAGNETIC_MODULE_V2 = "magneticModuleV2"
     THERMOCYCLER_MODULE_V1 = "thermocyclerModuleV1"
+    THERMOCYCLER_MODULE_V2 = "thermocyclerModuleV2"
```
Adding this enum value causes `thermocyclerModuleV2` to become visible in the HTTP API's schema and generated documentation. If this is acceptable, great, but if not, we may want to comment this out (as well as the downstream types that rely on it).

fsinapi replied:
Per discussion on Slack, I think we're okay including this at the level of the HTTP API.

mcous reviewed (commented):
Can smoke test this with a Gen1 TC on Monday to verify no breakage to existing flows. Answered @sfoster1's question about public interfaces below

sanni-t commented on api/src/opentrons/drivers/thermocycler/driver.py:
```diff
@@ -50,31 +49,91 @@ class GCODE(str, Enum):
+            ack=TC_GEN2_SERIAL_ACK,
+            retry_wait_time_seconds=0.1,
+            error_keyword="error",
+            alarm_keyword="alarm",
```
What is the 'alarm' keyword for?

fsinapi replied:
SerialConnection supports an AlarmResponse exception similar to the ErrorResponse exception, but with a different keyword. The `create` method defines default values that aren't duplicated by the `__init__` method, so this is just duplicating the behavior of constructing this object through `create`.

sanni-t commented on api/src/opentrons/drivers/thermocycler/driver.py:
```diff
@@ -50,31 +49,91 @@ class GCODE(str, Enum):
+            return ThermocyclerDriver(connection)
+
+    @staticmethod
+    async def is_gen2_protocol(connection: SerialConnection) -> bool:
```
can we rename this to something like `is_gen2_thermocycler` since that's what it's checking (and not whether a protocol uses a gen2 tc)?

sanni-t commented on shared-data/module/definitions/2/thermocyclerModuleV2.json:
```diff
@@ -0,0 +1,23 @@
+{
```
Not sure if it's necessary to add a schemaV2 definition.

fsinapi replied:
Sure, happy to remove. Is there any documentation (in this repo or elsewhere) on the reason for having these parallel schemaV2/V3 definitions?

sanni-t replied:
The V3 schema was added to accommodate heater-shaker module which doesn't fit into the V2 schema. These schemas aren't really meant to be used in parallel. We will probably switch the python api over to using V3 schemas quite soon and when that happens, having a TC Gen2 definition in V2 schema would just stay unused.
Also discussed this in a team meeting and we agreed that we'd prefer new definitions be added only to the latest schema being used. I'm going to do a quick test to see how involved it is to switch python api to V3 and get back to you.

sanni-t reviewed (approved):
Tested on a TC Gen1 and verified that this doesn't break any gen1 behavior. Just a couple of comments/ questions

fsinapi committed:
Removed schemaV2 def for Thermocycler Gen2
- Renamed `is_gen2_protocol` to `is_gen2_thermocycler` to avoid confusion with the use of the word `protocol`

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

fsinapi committed:
Merge branch 'edge' into api_add-tc_gen2-driver

# Conflicts:
#	api/src/opentrons/drivers/thermocycler/driver.py
#	api/src/opentrons/protocol_engine/types.py
#	api/tests/opentrons/protocol_engine/conftest.py

fsinapi committed:
Updated ThermocyclerModuleV2 def to match schema

sanni-t reviewed (approved):
Changes since my last review look good! Merge away!
